### ISSUE TITLE
JWK thumbprint?

### ISSUE THREAD

**Issue by FGasper** [DCIT/perl-CryptX#8]
As this library is something of a “grab-bag” of crypto functionality, would you consider incorporating JWK thumbprints, perhaps alongside export_key_jwk()?

```
    return _to_base64url(
        Digest::SHA::sha256(
            Cpanel::JSON::canonical_dump($self->_jwk_data())
        ),
    );
```

**Comment by karel-m** [comment]
Well, looking at https://tools.ietf.org/html/rfc7638 it will be a bit more complicated than what you have outlined bu basically it is doable.

According to rfc7638 it might be a good idea to add JWK thumbprint as a "kid" value into the exported JWK. I mean within the existing function `export_key_jwk()`.

**Comment by karel-m** [comment]
The easiest would be extending
```perl
sub export_key_jwk {
  my ($self, $type, $wanthash) = @_;
  ...
}
```
to
```perl
sub export_key_jwk {
  my ($self, $type, $wanthash, $wantkid) = @_;
  ...
  # put JWK thumbprint into "kid" value
}
```
But it'll make `export_key_jwk` even more ugly than it is now due to unusual `$wanthash` flag.

Or perhaps export `kid` value by default and introduce optional `$nokid` param.

**Comment by FGasper** [comment]
I was thinking a separate function, `export_key_jwk_thumbprint()`.

**Comment by karel-m** [comment]
For ECC it might look like this 6461c78 (untested, needs some tests + some documentation)

**Comment by FGasper** [comment]
Looks reasonable. My own immediate need is RSA rather than ECC, but it’ll be a simple port to do RSA.

**Comment by karel-m** [comment]
RSA is here https://github.com/DCIT/perl-CryptX/commit/92d87dc8dccb1b9b92673477d80cdcbf0c4ffb8c

Please test it + if you can please send a PR with some short doc + some added tests in `t/jwk.t'

**Comment by FGasper** [comment]
IMO this should allow both public and private thumbprints, with an interface similar to export_key_jwk().

```
sub export_key_jwk_thumbprint {
  my ($self, $type, $hash_name) = @_;
  $hash_name ||= 'SHA256';
  my $h = $self->export_key_jwk($type, 1);
  my $json = _encode_json({kty=>$h->{kty}, n=>$h->{n}, e=>$h->{e}});
  return digest_data_b64u($hash_name, $json);
}
```

**Comment by karel-m** [comment]
https://tools.ietf.org/html/rfc7638#section-3.2.1

> The JWK Thumbprint of a JWK representing a private key is computed as
   the JWK Thumbprint of a JWK representing the corresponding public
   key.  This has the intentional benefit that the same JWK Thumbprint
   value can be computed both by parties using either the public or
   private key.  The JWK Thumbprint can then be used to refer to both
   keys of the key pair.

**Comment by FGasper** [comment]
https://github.com/DCIT/perl-CryptX/pull/9

**Comment by karel-m** [comment]
I have just released CryptX-0.031 with this issue fixed
